src/platform/PICO/uart/uart_pio.c

```diff
@@ -0,0 +1,374 @@
+                        identifier, pinIndexMin, pinIndexMax);
+                continue;
+            } else {
+                uartPioBase = 16;
+            }
+        }
+
+        if (cfgRx) {
+            uartdev->rx = makePinDef(cfgRx);
+        }
+
+        if (cfgTx) {
+            uartdev->tx = makePinDef(cfgTx);
+        }
+
+        if (uartdev->rx.pin || uartdev->tx.pin ) {
+            bprintf("uartdev %p setting hardware to %p which has txbuffer %p", uartdev, hardware, hardware->txBuffer);
+            uartdev->hardware = (uartHardware_t *)hardware; // Sneak in pointer to pioUartHardware_t as a pointer to uartHardware_t
+        } else {
+            bprintf("** uartPinConfigure_pio no compatible rx or tx pin for this PIO UART");
+        }
+    }
+
+    bprintf("pico uartPinConfigure pio%d pin min, max = %d, %d; setting gpio base to %d", PIO_NUM(uartPio), pinIndexMin, pinIndexMax, uartPioBase);
+}
+
+static bool ensurePioProgram(PIO pio, const pio_program_t *program, bool isTx)
+{
+    // The GPIO base must be set before adding the program.
+    pio_set_gpio_base(uartPio, uartPioBase);
+
+    if (isTx) {
+        if (txProgram_offset < 0) {
+            txProgram_offset = pio_add_program(pio, program);
+        }
+
+        return txProgram_offset >= 0;
+   } else {
+        if (rxProgram_offset < 0) {
+            rxProgram_offset = pio_add_program(pio, program);
+        }
+
+        return rxProgram_offset >= 0;
+    }
+}
+
+#if SERIAL_PIOUART_COUNT > 0
+static void uartPioIrqHandler(uartPort_t *s, pioDetails_t *pioDetailsPtr)
+{
+    io_rw_32 *enableRegPtr = pioDetailsPtr->enableReg;
+    io_ro_32 *statusRegPtr = pioDetailsPtr->statusReg;
+    // Is RX enabled and does irq status indicate RX fifo not empty?
+    if (*enableRegPtr & pioDetailsPtr->rx_intr_bit && *statusRegPtr & pioDetailsPtr->rx_intr_bit) {
+        uint sm_rx = pioDetailsPtr->sm_rx;
+
+        // 8-bit read from the uppermost byte of the FIFO, as data is left-justified
+        io_ro_32 *rxfifo_shift = &uartPio->rxf[sm_rx];
+        serialReceiveCallbackPtr rxCallback = s->port.rxCallback;
+        if (rxCallback) {
+            void *rxCallbackData = s->port.rxCallbackData;
+            while (!pio_sm_is_rx_fifo_empty(uartPio, sm_rx)) {
+                const uint8_t ch = *rxfifo_shift >> 24; // left aligned, 8 bits
+                rxCallback(ch, rxCallbackData);
+            }
+        } else {
+            volatile uint8_t *rxBuffer = s->port.rxBuffer;
+            uint32_t rxBufferSize = s->port.rxBufferSize;
+            while (!pio_sm_is_rx_fifo_empty(uartPio, sm_rx)) {
+                const uint8_t ch = *rxfifo_shift >> 24; // left aligned, 8 bits
+                rxBuffer[s->port.rxBufferHead] = ch;
+                s->port.rxBufferHead = (s->port.rxBufferHead + 1) % rxBufferSize;
+            }
+        }
+    }
+
+    // Is TX enabled?
+    if (*enableRegPtr & pioDetailsPtr->tx_intr_bit) {
```

Comment:
Testing masked bit should work too - no need to check fifo without interrupt bit - space in fifo
```suggestion
    if (pioIntrBits & pioDetailsPtr->tx_intr_bit) {
```